spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java

```diff
@@ -307,32 +338,53 @@ public void deliveryComplete(IMqttToken token) {
 
 	}
 
+	@Override
+	public void connectComplete(boolean isReconnect) {
+		connectComplete(isReconnect, getUrl());
+	}
+
 	@Override
 	public void connectComplete(boolean reconnect, String serverURI) {
-		if (!reconnect) {
-			ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
-			String[] topics = getTopic();
-			this.topicLock.lock();
-			try {
-				if (topics.length > 0) {
-					int[] requestedQos = getQos();
-					this.mqttClient.subscribe(topics, requestedQos).waitForCompletion(getCompletionTimeout());
-					String message = "Connected and subscribed to " + Arrays.toString(topics);
-					logger.debug(message);
-					if (applicationEventPublisher != null) {
-						applicationEventPublisher.publishEvent(new MqttSubscribedEvent(this, message));
-					}
-				}
+		if (reconnect) {
+			return;
+		}
+		var clientManager = getClientManager();
+		if (clientManager != null && this.mqttClient == null) {
+			this.mqttClient = clientManager.getClient();
+		}
+
+		String[] topics = getTopic();
+		ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
+		this.topicLock.lock();
+		try {
+			if (topics.length == 0) {
+				return;
 			}
-			catch (MqttException ex) {
-				if (applicationEventPublisher != null) {
-					applicationEventPublisher.publishEvent(new MqttConnectionFailedEvent(this, ex));
-				}
-				logger.error(ex, () -> "Error subscribing to " + Arrays.toString(topics));
+
+			int[] requestedQos = getQos();
+			MqttSubscription[] subscriptions = IntStream.range(0, topics.length)
+					.mapToObj(i -> new MqttSubscription(topics[i], requestedQos[i]))
+					.toArray(MqttSubscription[]::new);
+			IMqttMessageListener listener = this::messageArrived;
+			IMqttMessageListener[] listeners = IntStream.range(0, topics.length)
+					.mapToObj(t -> listener)
+					.toArray(IMqttMessageListener[]::new);
+			this.mqttClient.subscribe(subscriptions, null, null, listeners, null)
+					.waitForCompletion(getCompletionTimeout());
```

Comment:
Don't we need to do the same `subscribeToken.getGrantedQos()` logic over here as well, like you have introduced in v3 impl?

Reply:
For v5 we didn't have it initially. So I am not sure. Probably not?

Reply:
Hm, I don't find this `0x80` in the blocking client in v5 as well. Probably it was something related to v3 only.